controller,worker: Add garbage collection to delete old slug releases

lmars opened the pull request:
This adds app garbage collection to remove old slug releases based on the value of `app.Meta["gc.max_inactive_slug_releases"]`:

* if not set or set to `false` then do nothing
* if set to an integer, ensure the app only has that many releases with distinct slugs

Garbage collection cycles are scheduled after every deployment.

For now I have left it as opt in so we can test it out in the real world and then potentially make it the default in the future.

To opt in, run:

```
$ flynn meta set gc.max_inactive_slug_releases=N
```

where `N` is a positive integer.

Ref #2750.

titanous commented:
Does this take into account the fact that multiple apps can reference the same release? We should only garbage collect if the maximum GC parameter across all referencing releases allows it.

lmars commented:
@titanous releases cannot be deleted if they are currently referenced as an app release (enforced by [this trigger](https://github.com/flynn/flynn/blob/master/controller/schema.go#L351-L361) in the database) but yes there is a chance that a given release could be within the last N of another app, but there is no reliable way to determine the last N releases for a given app, we are currently just relying on the fact that slug releases are most likely created in order for a single app.

I do wonder whether we need the flexibility of releases belonging to multiple apps, the situation would be simpler if releases belong to a single app and we just copy them if we want them to relate to another app. WDYT?

titanous commented:
Yeah, copying probably makes sense. There should be few actual instances of this, as it's only available via the API.

titanous commented:
It looks like this is currently just called generic "garbage collection" throughout the type names, but is scoped to releases/artifacts of a single app. Should we call it "App Garbage Collection" or "App Release GC" or similar? We may have GC for other things in the controller in the future.

lmars commented:
@titanous I renamed things to "App Garbage Collection" and I also changed release deletion so that it is namespaced by app and only deletes the release if there are no formations for other apps.

I've made a breaking change to the controller API (moving `DELETE /releases/:release_id` to `DELETE /apps/:app_id/releases/:release_id`), but given we only added the route in the last week or so and it is not in a stable release, I don't think we need to support backwards compatibility.

titanous commented:
Does this ignore releases with formations currently deployed? Should the parameter be called `gc.max_inactive_slug_releases`?

lmars commented:
@titanous 

> Does this ignore releases with formations currently deployed? Should the parameter be called gc.max_inactive_slug_releases?

It did not, so I have updated the worker to only delete scaled down releases and renamed the metadata to `gc.max_inactive_slug_releases` as suggested.

I have also updated release deletion so that artifacts are only deleted once all associated releases have been deleted.

The test has been updated to handle the above two cases.

titanous commented:
LGTM

dottodot commented:
Can you clarify if this will help with the ever increasing disk use issue. Also I'm still unsure how to tell when something has been added to the stable branch or not.

lmars commented:
@dottodot 

> Can you clarify if this will help with the ever increasing disk use issue

It should do yes.

> I'm still unsure how to tell when something has been added to the stable branch or not

Yes this is currently not very well exposed, I will open an issue to discuss approaches for making this information more available.

This particular change is not yet in stable (currently `v20160512.0`) but it did appear in nightly from version `v20160515.0`.

dottodot commented:
Ok might have to give the nightly a go then as my only issue with flynn at the moment is I have to build a new cluster every week as it keep running out of space when pushing lots of updates.

lmars commented:
@dottodot with the latest stable version, you can actually delete old releases, it is just a manual process using `flynn release delete <id>` (rather than being done for you by the changes merged here).

dottodot commented:
Ah brillant, hadn't realised that.
